internal/worker/bootstrap/addressfinder.go

```diff
@@ -75,3 +75,42 @@ func BootstrapAddressFinder(
 		return getInstanceAddresses(ctx, lister, bootstrapInstance)
 	}
 }
+
+// BootstrapCAASAddressFinder is responsible for finding the bootstrap network
+// addresses for a bootstrap instance.
+func BootstrapCAASAddressFinder(
+	providerFactory providertracker.ProviderFactory, namespace string,
+	// providerGetter providertracker.ProviderGetter[caas.ServiceManager],
+) BootstrapAddressFinderFunc {
+	return func(
+		ctx context.Context,
+		bootstrapInstance instance.Id,
+	) (network.ProviderAddresses, error) {
+		providerGetter := providertracker.ProviderRunner[caas.ServiceManager](
+			providerFactory, namespace,
+		)
+
+		svcManager, err := providerGetter(ctx)
+		if err != nil {
+			return network.ProviderAddresses{}, fmt.Errorf(
+				"cannot get service manager from provider for finding bootstrap addresses: %w",
+				err,
+			)
+		}
+
+		// Retrieve the k8s service from the k8s broker.
+		svc, err := svcManager.GetService(ctx, k8sconstants.JujuControllerStackName, true)
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+		if svc == nil || len(svc.Addresses) == 0 {
+			// If no addresses are returned from the K8s broker, we return
+			// "localhost" as a hard-coded address, this guarantees that the
+			// bootstrap instance will be able to connect to the controller
+			// service.
+			return network.NewMachineAddresses([]string{"localhost"}).AsProviderAddresses(), nil
```

Comment:
We have modelled host names separately to IP addresses in Dqlite. This should be the loopback _address_.